torchtune/models/qwen2_5/_tokenizer.py

```diff
@@ -0,0 +1,232 @@
+            word, e.g. Chinese); technically not a memory leak but appears as one.
+            By default, we set the cache size equals to size of the official Qwen2 tokenizer.
+
+    Example:
+        >>> tokenizer = Qwen2Tokenizer(
+                path="/path/to/vocab.json", merges_file="/path/to/merges.txt", special_tokens=QWEN2_SPECIAL_TOKENS)
+        >>> tokenized_text = tokenizer.encode("Hello world!")
+        >>> print(tokenized_text)
+        [39, 385, 78, 675, 0, 2000]
+    """
+
+    def __init__(
+        self,
+        path: str,
+        merges_file: str,
+        special_tokens: Dict[str, int] = QWEN2_5_SPECIAL_TOKENS,
+        max_seq_len: Optional[int] = None,
+        *,
+        prompt_template: Optional[PromptTemplate] = None,
+        errors: str = "replace",
+        unk_token: Optional[str] = ENDOFTEXT,
+        bos_token: Optional[str] = None,
+        eos_token: str = ENDOFTEXT,
+        pad_token: Optional[str] = ENDOFTEXT,
+        bpe_cache_size: int = DEFAULT_QWEN2_TOKENIZER_BPE_CACHE_SIZE,
+    ):
+        super().__init__(
+            path=path,
+            merges_file=merges_file,
+            special_tokens=special_tokens,
+            max_seq_len=max_seq_len,
+            prompt_template=prompt_template,
+            errors=errors,
+            unk_token=unk_token,
+            bos_token=bos_token,
+            eos_token=eos_token,
+            pad_token=pad_token,
+            bpe_cache_size=bpe_cache_size,
+        )
+
+    def tokenize_messages(
+        self,
+        messages: List[Message],
+        *,
+        add_eos: bool = True,
+    ) -> Tuple[List[int], List[bool]]:
+        """
+        Given a list of messages, return a list of tokens for the concatenated
+        and formatted messages.
+
+        Args:
+            messages (List[Message]): The message list to tokenize.
+            add_eos (bool): Wether to add the tokenizer's eos_id at the end of the
+                sequence of messages. Default is True.
+
+        Returns:
+            Tuple[List[int], List[bool]]: The list of token ids and the list of masks.
+
+        Raises:
+            RuntimeError: If a message contains non-text content
+        """
+        assert not isinstance(
+            self.prompt_template, ChatMLTemplate
+        ), "Using ChatMLTemplate with tokenize_messages will result in multiple <|im_*|> tokens wrapping each message."
+        templated_messages = (
+            self.prompt_template(messages)
+            if self.prompt_template is not None
+            else messages
+        )
+
+        tokenized_messages = []
+        mask = []
+        for i, message in enumerate(templated_messages):
+            tokens = []
+
+            # message header
+            if message.role == "ipython":
+                if i == 0 or templated_messages[i - 1].role != "ipython":
+                    self._add_message_start_tokens(tokens, "user")
```

Comment:
do you mind adding a comment here explaining why we're adding a user header and checking the previous message? not sure what cases this is covering